proposals/simd-0182-conditional-cu-metering.md

```diff
@@ -0,0 +1,114 @@
+CUs.
+
+## Motivation
+
+In the Solana protocol, tracking transaction Compute Unit (CU) consumption is a
+critical aspect of maintaining consensus. Block costs are part of this
+consensus, meaning that all clients must agree on the execution cost of each
+transaction, including those that error out during execution. Ensuring
+consistency in CU tracking across clients is essential for maintaining protocol
+integrity.
+
+To improve performance, Solana programs are often compiled with a JIT that works
+at the level of Basic Blocks — linear sequences of sBPF instructions with a
+single entry and exit point, and no loops or branches. Basic Blocks allow for
+efficient execution by reducing the overhead associated with tracking CU
+consumption for each individual sBPF instruction.
+
+Other than in rare, exceptional situations discussed below, the total CU
+consumption for a Basic Block is deterministic and CU accounting can be
+done once per basic block instead of at each instruction. A transaction
+completing successfully or with most errors implies that execution exited each
+basic block at its single exit point, and thus that the total CU consumption of
+the execution is equal to the sum of the CU cost of each Basic Block executed.
+
+However, when an exception is thrown during the execution of a Basic Block
+(e.g., a null memory dereference or other faults), determining the exact number
+of CUs consumed up to the point of failure requires additional effort. For
+instance, the Agave client implements a mechanism that tracks the Instruction
+Pointer (IP) or Program Counter (PC) to backtrack and estimate the CUs consumed
+when an exception occurs. More details on this mechanism can be found
+[here](https://github.com/solana-labs/rbpf/blob/57139e9e1fca4f01155f7d99bc55cdcc25b0bc04/src/jit.rs#L267).
+
+While this approach is effective, it introduces additional work and complexity.
+These mechanisms are often implementation-specific, and requiring all clients to
+track the exact number of executed sBPF instructions for consensus is costly and
+unnecessary. Such precision is not essential for protocol-level consensus,
+especially since these cases are rare.
+
+Instead of mandating implementation-specific work to handle exceptions, we
+propose the following clarification in the protocol:
+
+- For successful execution of a Basic Block (i.e., the block exits at the last
+ sBPF instruction), the deterministic CU cost of the block will be charged to
+the transaction’s CU meter. This ensures that CU consumption for successful
+transactions is accurately accounted for.
+- In the event of irregular failure, where execution aborts from the middle of
+basic block, the requested CUs for the transaction will be charged to the CU
+meter. This allows for a simple and efficient fallback mechanism that avoids the
+need for tracking the exact number of executed instructions up to the point of
+failure.
```

Comment:
What exactly constitutes an "irregular failure"? If we go down this road, distinguishing between regular and irregular errors could end up being even more complex than the current implementation.

Reply:
Noticing you define it again under terminology, but I'm not so sure the rest of the VM error codes are considered "irregular". This could be a semantic argument though.